src/celeritas/optical/ScintillationGenerator.hh

```diff
@@ -0,0 +1,216 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2024 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/optical/ScintillationGenerator.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "corecel/Assert.hh"
+#include "corecel/Macros.hh"
+#include "corecel/Types.hh"
+#include "corecel/data/Collection.hh"
+#include "corecel/math/ArrayOperators.hh"
+#include "corecel/math/ArrayUtils.hh"
+#include "celeritas/random/distribution/BernoulliDistribution.hh"
+#include "celeritas/random/distribution/ExponentialDistribution.hh"
+#include "celeritas/random/distribution/GenerateCanonical.hh"
+#include "celeritas/random/distribution/NormalDistribution.hh"
+#include "celeritas/random/distribution/UniformRealDistribution.hh"
+
+#include "OpticalDistributionData.hh"
+#include "OpticalPrimary.hh"
+#include "ScintillationData.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Sample Scintillation photons.
+ *
+ * \note This performs the same sampling routine as in G4Scintillation class
+ * of the Geant4 release 11.2 with some modifications.
+ */
+class ScintillationGenerator
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using Energy = units::MevEnergy;
+    //!@}
+
+  public:
+    // Construct from scintillation data and distribution parameters
+    inline CELER_FUNCTION
+    ScintillationGenerator(OpticalDistributionData const& dist,
+                           NativeCRef<ScintillationData> const& shared,
+                           Span<OpticalPrimary> photons);
+
+    // Sample Scintillation photons from the distribution
+    template<class Generator>
+    inline CELER_FUNCTION Span<OpticalPrimary> operator()(Generator& rng);
+
+    // TODO: Move it to detail/OpticalUtils.
+    static CELER_CONSTEXPR_FUNCTION real_type hc()
+    {
+        return constants::h_planck * constants::c_light;
+    }
```

Comment:
This seems too simple to move to optical utils :) I think it'd be better inlined into the code (a single operation doesn't seem worth creating a static function to help eliminate).

Reply:
Okay, moved it to the code.